# Cross_val_score
This is the **right way** to use cross validation

In [66]:
from sklearn import datasets
from sklearn.cross_validation import cross_val_score
iris= datasets.load_iris()
X=iris.data
y=iris.target


In [67]:
from sklearn.svm import SVC

In [68]:
clf = SVC(kernel='linear', C=1)

In [69]:
scores = cross_val_score(clf, X, y, cv=6)

In [71]:
scores

array([ 0.96296296,  1.        ,  0.95833333,  0.95833333,  1.        ,  1.        ])

In [72]:
import numpy as np
np.mean(scores)   # This the mean score for mean cross validation error where K=5 

0.97993827160493829

# What is the optimum number of folds? 

In [74]:
from sklearn import datasets
from sklearn.cross_validation import cross_val_score
iris= datasets.load_iris()
X=iris.data
y=iris.target

from sklearn.svm import SVC
import numpy as np 

success=[]
N= range(2,20)
for k in N:
    clf = SVC(kernel = 'linear', C=1)
    scores=cross_val_score(clf, X, y, cv=k)
    success.append(np.mean(scores))
    

    

In [77]:
success

[0.96666666666666667,
 0.97998366013071891,
 0.97970085470085477,
 0.98000000000000009,
 0.97993827160493829,
 0.97363945578231292,
 0.9732142857142857,
 0.98641975308641983,
 0.97333333333333338,
 0.97424242424242424,
 0.97361111111111109,
 0.97435897435897434,
 0.97420634920634919,
 0.97222222222222221,
 0.97395833333333326,
 0.97385620915032689,
 0.97530864197530875,
 0.97660818713450304]

In [76]:
len(success)

18

In [78]:
max(success)

0.98641975308641983

As we can see, the highest cross validation score is 0.98641975308641983, when **k=9**

# Using KFold to get the cross value score

In [82]:
from sklearn.cross_validation import KFold
from sklearn.svm import SVC

In [93]:
kf = KFold(len(X), n_folds=4, shuffle=True)

In [99]:
scores=[]
for train_index, test_index in kf:
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]
    clf= SVC(kernel='linear', C=1.)
    clf.fit(X_train, y_train)
    scores.append(clf.score(X_test, y_test))

In [100]:
scores

[0.97368421052631582, 0.92105263157894735, 1.0, 0.97297297297297303]

In [101]:
import numpy as np 
np.mean(scores)

0.96692745376955913